{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b44fee27-08ad-4f6f-b59f-68344e2ce118",
   "metadata": {},
   "source": [
    "# **COUNT MATRIX CREATION**\n",
    "## Alignment files are concatenated into a matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "40f78137-e4d5-49fc-b401-2f209e59990c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "from numpy import isneginf\n",
    "import pandas as pd\n",
    "import glob\n",
    "import sys\n",
    "import csv\n",
    "import re"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0e658c4-a357-4322-9fdb-571034160f23",
   "metadata": {},
   "source": [
    "## Load metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4b94e452-cf1f-4bb2-8d55-081f53ce4bae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Condition</th>\n",
       "      <th>Replicate</th>\n",
       "      <th>Filename_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0h_1</td>\n",
       "      <td>0h</td>\n",
       "      <td>1</td>\n",
       "      <td>A1_S2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0h_2</td>\n",
       "      <td>0h</td>\n",
       "      <td>2</td>\n",
       "      <td>A2_S8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0h_3</td>\n",
       "      <td>0h</td>\n",
       "      <td>3</td>\n",
       "      <td>A3_S9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0h-Epi_1</td>\n",
       "      <td>0h-Epi</td>\n",
       "      <td>1</td>\n",
       "      <td>A5_S29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0h-Epi_2</td>\n",
       "      <td>0h-Epi</td>\n",
       "      <td>2</td>\n",
       "      <td>A6_S30</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Sample Condition  Replicate Filename_ID\n",
       "0      0h_1        0h          1       A1_S2\n",
       "1      0h_2        0h          2       A2_S8\n",
       "2      0h_3        0h          3       A3_S9\n",
       "3  0h-Epi_1    0h-Epi          1      A5_S29\n",
       "4  0h-Epi_2    0h-Epi          2      A6_S30"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta=pd.read_csv('data/gastruloid_bulk_metadata.csv')\n",
    "meta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61a991ed-a2f0-4581-92ed-f6310ec5556f",
   "metadata": {},
   "source": [
    "## List gene counts files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4b9d4f58-bc93-4f68-a110-94693689d03a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "path = 'star_out'\n",
    "\n",
    "files = glob.glob(os.path.join(path , \"*_S*/*_ReadsPerGene.out.tab\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "dd5fa1d2-ff41-4985-a9fa-edb571cb08b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# retain only files specified in metadata\n",
    "files=[f for f in files if f.split('/')[1] in meta['Filename_ID'].str.split('_R1').str[0].values]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b8a6ba75-0737-44fd-ae9e-8606329d12bc",
   "metadata": {},
   "source": [
    "## Sort files according to metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4351d0b4-63cc-423d-9039-53fe999dd0b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# set order using metadata\n",
    "order=[x.split('_R1')[0] for x in meta['Filename_ID']]\n",
    "# sort files\n",
    "sorted_files=sorted(files, key=lambda x: order.index(x.split('/')[1]))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21cb0c47-3831-4050-be1e-cd684acf30b1",
   "metadata": {},
   "source": [
    "## Countmatrix: read and concatenate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8a3f0cc4-c8ed-4c4b-a5c8-79f4db8951fc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Sample</th>\n",
       "      <th>0h_1</th>\n",
       "      <th>0h_2</th>\n",
       "      <th>0h_3</th>\n",
       "      <th>0h-Epi_1</th>\n",
       "      <th>0h-Epi_2</th>\n",
       "      <th>0h-Epi_3</th>\n",
       "      <th>48h_1</th>\n",
       "      <th>48h_2</th>\n",
       "      <th>48h_3</th>\n",
       "      <th>48h-AF_1</th>\n",
       "      <th>...</th>\n",
       "      <th>120h-high-ActA_1</th>\n",
       "      <th>120h-high-ActA_2</th>\n",
       "      <th>120h-high-ActA_3</th>\n",
       "      <th>144h-low-ActA_1</th>\n",
       "      <th>144h-low-ActA_2</th>\n",
       "      <th>144h-low-ActA_3</th>\n",
       "      <th>144h-high-ActA_1</th>\n",
       "      <th>144h-high-ActA_2</th>\n",
       "      <th>144h-high-ActA_3</th>\n",
       "      <th>96h-3μMc-DMSO_1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Xkr4</th>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>38</td>\n",
       "      <td>55</td>\n",
       "      <td>41</td>\n",
       "      <td>9</td>\n",
       "      <td>6</td>\n",
       "      <td>17</td>\n",
       "      <td>16</td>\n",
       "      <td>...</td>\n",
       "      <td>27</td>\n",
       "      <td>43</td>\n",
       "      <td>19</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>21</td>\n",
       "      <td>7</td>\n",
       "      <td>16</td>\n",
       "      <td>46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gm18956</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gm37180</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td>...</td>\n",
       "      <td>9</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gm37363</th>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>22</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>9</td>\n",
       "      <td>...</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gm37686</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 68 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Sample   0h_1  0h_2  0h_3  0h-Epi_1  0h-Epi_2  0h-Epi_3  48h_1  48h_2  48h_3  \\\n",
       "Gene                                                                           \n",
       "Xkr4        0     3     0        38        55        41      9      6     17   \n",
       "Gm18956     0     0     0         0         0         0      0      0      0   \n",
       "Gm37180     0     0     0         4         3         1      2      0      2   \n",
       "Gm37363     0     4     0        11        22         1      0      1      8   \n",
       "Gm37686     0     0     7         0         9         0      2      1      3   \n",
       "\n",
       "Sample   48h-AF_1  ...  120h-high-ActA_1  120h-high-ActA_2  120h-high-ActA_3  \\\n",
       "Gene               ...                                                         \n",
       "Xkr4           16  ...                27                43                19   \n",
       "Gm18956         0  ...                 0                 0                 2   \n",
       "Gm37180         7  ...                 9                 5                 0   \n",
       "Gm37363         9  ...                11                11                11   \n",
       "Gm37686         0  ...                 0                 0                 3   \n",
       "\n",
       "Sample   144h-low-ActA_1  144h-low-ActA_2  144h-low-ActA_3  144h-high-ActA_1  \\\n",
       "Gene                                                                           \n",
       "Xkr4                   5                0                9                21   \n",
       "Gm18956                0                0                0                 0   \n",
       "Gm37180                0                2                0                 1   \n",
       "Gm37363                0                5                4                 0   \n",
       "Gm37686                0                0                0                 0   \n",
       "\n",
       "Sample   144h-high-ActA_2  144h-high-ActA_3  96h-3μMc-DMSO_1  \n",
       "Gene                                                          \n",
       "Xkr4                    7                16               46  \n",
       "Gm18956                 3                 0                0  \n",
       "Gm37180                 0                 0                6  \n",
       "Gm37363                 0                 5               11  \n",
       "Gm37686                 3                 5                5  \n",
       "\n",
       "[5 rows x 68 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs = []\n",
    "samples =[]\n",
    "\n",
    "for filename in sorted_files:\n",
    "    df = pd.read_csv(filename, index_col=0, header=0, sep='\\t', skiprows=3)\n",
    "    df = df.iloc[:,0] # first column belongs to Unstranded reads (sequencing specififcation)\n",
    "\n",
    "    dfs.append(df)\n",
    "    \n",
    "# concat dfs\n",
    "counts = pd.concat(dfs, axis=1, ignore_index=True)\n",
    "# remove genes with 0 counts accross all genes\n",
    "counts = counts.loc[~(counts==0).all(axis=1)]\n",
    "# rename columns and index\n",
    "counts.columns = meta['Sample']\n",
    "counts.index.name = 'Gene'\n",
    "counts = counts.fillna(0)\n",
    "\n",
    "# save df\n",
    "counts.to_csv(path + '/gastruloid_bulk.csv')\n",
    "counts.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
